### ISSUE TITLE
Typelevel discord feedback

### ISSUE THREAD

**Issue by jisantuc** [pcaisse/regex-parser#1]
I shared this repo in Typelevel Discord and got the following feedback --

> This looks really great! We should have a link to examples in the cats-parse README and this would be great. My main feedback would be that I always recommend testing parsers with scalacheck. Here you could add a serialize method to your RegExp, and then use scalacheck to generate random regexps, call serialize, make sure they parse identically.
two other nits: it looks like the main parser is private, and only the method is exposed... a bit hard to follow due to omitted return types. Secondly, I think not having all the Regexps extends a single type is not a great choice. Using unions to make more precise types is fine, but I think they should share a common top type, such as RegExp or Regex

So it sounds like there are three pieces of feedback here:

- [x] round trip tests with Scalacheck would be nice
- [x] add a return type to the exposed `parse` method
- [x] make the Regexps share a type

**Comment by pcaisse** [comment]
Donezo: https://github.com/pcaisse/regex-parser/compare/d443ec4d0b4e82f051f52c5339f5ced85e770165...HEAD
